Scorecard article (#15)
balteravishay/scorecard: !article/article.md

![pinned-dependencies](./pinned-dependencies.png)
![packaging](./packaging.png)
 
Other checks illustrated below, which are not tied to specific ecosystems and rely solely on GitHub APIs, highlight low-hanging fruit opportunities for .NET ecosystem maintainers to make improvements and enhance open-source security.
 
 ----add rest of the images
  
# Recommendations and action items
With the recent improvements in handling pinned dependencies and packaging checks for the NuGet package manager, the ability to assess the security features of GitHub repositories in the .NET ecosystem will be substantially enhanced. This should encourage repositories’ maintainers to adopt necessary improvements, leading to higher scores measured by Scorecard and improved overall security.

In conclusion, regularly running OSSF Scorecard checks helps ensure your project stays secure, up-to-date, and aligned with coding best practices. This proactive approach significantly reduces the risk of security vulnerabilities within your software ecosystem.

> Don’t delay! Check the Scorecard of your favorite GitHub repository today and if you’d like to improve the score, take a look at the suggested mitigation steps 😊

Reviewer: ```suggestion
Don’t delay! Check the Scorecard of your favorite GitHub repository today and if you’d like to improve the score, take a look at the suggested mitigation steps 😊.
```

# Introduction

> OpenSSF Scorecard is a tool developed by the Open Source Security Foundation (OpenSSF) that provides automated security assessments for open-source projects. The primary goal of the Scorecard project is to help developers and users determine the security posture of open-source software by generating a "scorecard" based on a series of security-related checks. 

Reviewer 2: ```suggestion
OpenSSF Scorecard is a tool developed by the Open Source Security Foundation (OpenSSF) that provides automated security assessments for open-source projects. The primary goal of the Scorecard project is to help developers and users determine the security posture of open-source software by generating a "score" based on a series of security-related checks. 
```

Using Scorecard, developers can assess the risks associated with each check and make informed decisions about accepting these risks, evaluating alternative solutions, or making improvements. 

> Additionally, a company’s Open Source Program Office (OSPO) can utilize the tool at an organizational level to assess the scores of their owned repositories, as is the case with the Microsoft OSPO.

Reviewer 2: TODO: check with OSPO if we can publish this

Author: Resolving this as we clarified.

The image below illustrates the score of a Microsoft owned repository.

--- add image
 
# Content
This article covers:

> -	TODO: Once we agree on a final form add a paragraph with what we will cover in the article

Reviewer 2: suggest something like: 

This post is tailored specifically for .Net maintainers, highlighting how Scorecard and the newly introduced support for .Net can help enhance the security and quality of the ecosystem. It will explore what this support means for you, what are the gaps that you can help us build next into Scorecard and dives into our research of the .NET open-source ecosystem to uncover several low-hanging fruits that maintainers should address to improve their project's security posture. We believe that by focusing on these areas, we can bring a significant enhancement to our community's security posture and a safer open-source ecosystem

Author: Excellently written, thank you!

2.	Scoring System: Each project is evaluated against these security checks and receives a score (on a scale of 0 to 10) that reflects the overall security maturity of the project.
3.	Security Best Practices: The checks are based on security best practices for open-source development, such as the use of pinned dependencies, branch protection, code reviews, and fuzzing.
4.	Ease of Use: Developers and users can easily run Scorecard checks on their repositories via GitHub Actions, as well as view public scorecards for popular open-source projects.
5.	Community and Ecosystem Support: As part of OpenSSF’s initiatives, the Scorecard project helps improve the overall security of the open-source ecosystem by encouraging maintainers to adopt secure development practices.

Using the OSSF Scorecard for a project is important because it helps ensure that the open-source software you maintain follows essential security best practices. It can also be used to evaluate whether dependencies you take into your codebase meet your expected level of adherence to security practices. 

> Today, Scorecard is recognized and utilized across various sectors of the software engineering industry. For instance, it is mentioned in the Cybersecurity and Infrastructure Security Agency (CISA) Tools and Resources (OpenSSF Scorecard | CISA). Moreover, Sonatype, who specialize in software supply chain automation and security, references its checks and recommends its adoption in their annual reports. Google’s Open Source Insights (deps.dev) project highlights the scorecard checks of each package shown.

Reviewer 2: missing some links that existed in the word version

Author: I will add them, thanks!

### 6. Encouraging Secure Best Practices
Scorecard encourages project maintainers to adopt secure practices like branch protection, regular security reviews, and integrating automated tools in CI/CD pipelines. By using Scorecard, teams can identify and improve weak areas, ensuring they consistently follow security best practices as their project evolves.

### 7. Preventing Supply Chain Attacks
A major concern in modern software development is the risk of supply chain attacks, where attackers compromise project dependencies or processes to inject malicious code. Scorecard assesses how effectively a project manages its dependencies, which is vital for preventing such attacks. Projects with strong scores in dependency management ensure that their dependencies are updated, verified, and secure, minimizing the risk of exploitation by malicious actors.

# What are the checks run by Scorecard 
The checks from the following table are all run against the target project by default. 

> Except for Pinned-Dependencies and Packaging checks, that are ecosystem-specific, the rest of the checks are run with the same criteria on every repository, regardless of the language used in the repo.

Reviewer 2: i actually think that adding the "ecosystem dependency" sentence and the column in the table is a bit confusing since we see that there are ecosystem dependencies in support for features that we didn't think there would be such (i.e. works for GH but not for ADO, works for shell but not for powershell, etc.) wdyt about removing it?

Author: Yes, I agree. It is also more concise without.

For example: `--nuget=Newtonsoft.Json`

Note that this feature is only supported for packages that define the repository attribute in their  package metadata .nuspec file.

## Packaging 
This check evaluates whether the project is published as a package. Packages make it easier for users to download, install, update, and receive security patches via a package manager. The check currently looks for GitHub packaging workflows and language-specific GitHub Actions that upload the package to NuGet and other package managers. 

Note that if a project uses different packaging tools, it may still receive a low score, as Scorecard can't detect all packaging methods. A low score doesn’t necessarily mean the project is at risk. If your packaging method isn’t recognized, consider opening an issue with the Scorecard maintainers. 

In case your project receives a low score on this check, you can try the following remediation steps:

- Publish your project as a downloadable package, e.g., if hosted on GitHub, use GitHub's mechanisms for publishing a package.
- If hosted on GitHub, use a GitHub action to release your package to language-specific hubs.

## Pinned dependencies 

> For restoring NuGet packages, it is considered best practice to use Central Package Management (CPM) and/or a lock file to fully lock down the package dependency graph, ensuring consistent and repeatable builds while preventing “it works on my machine" issues.

Reviewer 2: perhaps more importantly than preventing "works on my machine" issues is [counter dependency confusion (aka substitution) attacks](https://azure.microsoft.com/en-us/resources/3-ways-to-mitigate-risk-using-private-package-feeds/). wdyt?

Author: For some reason, I cannot open the link, it redirects me to the homepage.

Reviewer 2: sorry, wrong link: https://learn.microsoft.com/en-us/nuget/concepts/security-best-practices#nuget-feeds

- For Dockerfiles, pin dependencies by hash as in this example. If using a manifest list, pin to the manifest list hash instead.
- In GitHub workflows used in building and releasing your project, pin dependencies by hash. See main.yaml for example. To determine the permissions needed for your workflows and pin actions to commit SHAs, you may use StepSecurity's online tool.
- To update dependencies after pinning, use tools such as Dependabot and Renovate bot.

# An analysis of Scorecard checks for .NET open source repositories on GitHub

On GitHub, there are over 1,000 C# and F# repositories with more than 1,500 stars. Our team analyzed the Scorecard scores for these repositories using the public Scorecard API to get an ecosystem-wide view of the state of .NET open-source projects. 

The diagram below displays the average scores for all Scorecard checks for those repositories. Please note that scores below zero are not relevant and may indicate an error during the Scorecard check execution.

![average](./average-scores.png)

> It’s especially noteworthy to look at the most critical security check, Dangerous Workflows. In this case, the check either receives a perfect score of 10 or identifies an error, which occurs in over 250 repositories. This reveals existing security gaps and highlights opportunities for improvement to strengthen overall security.

Reviewer 2: is this a security gap or a support gap that highlights an opportunity to improve by contributing?

Author: Support gap, I reworded it.

![dangerous-workflow](./dangerous-workflow.png)
 
Moreover, the pinned dependency scores show a gap in support for the NuGet ecosystem that should be improved given the recent work that was done. As these features were newly released at the time this article was written, the change in score is still not visible. There is a similar gap in support for packaging that can be better addressed by Scorecard to further detect the NuGet ecosystem.

![pinned-dependencies](./pinned-dependencies.png)
![packaging](./packaging.png)

 

> Other checks illustrated below, which are not tied to specific ecosystems and rely solely on GitHub APIs, highlight low-hanging fruit opportunities for .NET ecosystem maintainers to make improvements and enhance open-source security.

Reviewer 2: is it worth mentioning what specifically are the low hanging fruits here?

Author: Here we are actually missing some images which show the remaining checks. I did not add them yet because I was waiting on the higher resolution images.
I can also mention them in a paragraph, after I add the pictures.

 
 ----add rest of the images

  
# Recommendations and action items
With the recent improvements in handling pinned dependencies and packaging checks for the NuGet package manager, the ability to assess the security features of GitHub repositories in the .NET ecosystem will be substantially enhanced. This should encourage repositories’ maintainers to adopt necessary improvements, leading to higher scores measured by Scorecard and improved overall security.

In conclusion, regularly running OSSF Scorecard checks helps ensure your project stays secure, up-to-date, and aligned with coding best practices. This proactive approach significantly reduces the risk of security vulnerabilities within your software ecosystem.

> Don’t delay! Check the Scorecard of your favorite GitHub repository today and if you’d like to improve the score, take a look at the suggested mitigation steps 😊

Reviewer 2: I would also add a list of open scorecard issues that folks are invited to collaborate on, wdyt? (we should first open them of course :))

Author: I added some of them, let me know if I missed anything.

# Introduction
[OpenSSF Scorecard](https://github.com/ossf/scorecard) is a tool developed by the Open Source Security Foundation (OpenSSF) that provides automated security assessments for open-source projects. The primary goal of the Scorecard project is to help developers and users determine the security posture of open-source software by generating a **score** based on a series of security-related checks. 

> ![nuget](./Scorecard-NuGet.png)

Reviewer: @Ioana37 this image is rendering very big, would it possible to have it smaller?

Using Scorecard, developers can assess the risks associated with each check and make informed decisions about accepting these risks, evaluating alternative solutions, or making improvements. 
Additionally, a company’s Open Source Program Office (OSPO) can [utilize the tool](https://openssf.org/blog/2022/09/29/how-ospos-can-be-a-key-lever-for-open-source-sustainability-and-security/) at an organizational level to assess the scores of their owned repositories, as is the case with the Microsoft OSPO.

> The image below illustrates the score of a Microsoft owned repository.

Reviewer 2: @Ioana37 we have the green light for this. here is a suggestion that OSPO have made to the phrasing: 

"While projects can use GitHub Actions to generate a scorecard for their specific project (the OSSF also generates scorecards on behalf of many popular open source projects), it's also possible for a security team or an Open Source Programs Office to generate the scorecards at scale. For example, at Microsoft, OSPO regularly computes scores for all public and private repositories, making historical information and trends available to teams."

also: 

Maybe also, so people think it's not just a score, "Project owners can dig into all aspects of the scorecard." - as part of the ALT text for the image? Right now it might look to some like it's just a score, but it's more.

Author: Added the suggested paragraph and the clarification.

SAST | Does the project use static code analysis tools, e.g. CodeQL, LGTM (deprecated), SonarCloud? | Medium
Security-Policy | Does the project contain a security policy? | Medium
Signed-Releases | Does the project cryptographically sign releases? | High
Token-Permissions | Does the project declare GitHub workflow tokens as read only? | High
Vulnerabilities | Does the project have unfixed vulnerabilities? Uses the OSV service. | High
Webhooks | Does the webhook defined in the repository have a token configured to authenticate the origins of requests? | Critical

# What is supported for .NET/NuGet
For the past two years, Scorecard has started to implement dedicated support for projects within the .NET ecosystem, which use the NuGet package manager. The most recent features added are support for checking pinned dependencies when restoring packages using a [lock file](https://devblogs.microsoft.com/nuget/enable-repeatable-package-restores-using-a-lock-file/#how-to-enable-the-lock-file) and when using [Central Package Management](https://learn.microsoft.com/en-us/nuget/consume-packages/Central-Package-Management).

> 

Reviewer 2: ```suggestion

We would like to extend a special Thank You for the team who contributes and coordinates these contributions:

[Avishay Balter](https://www.linkedin.com/in/avishay-balter/)
[Ioana Amarande](https://www.linkedin.com/in/ioana-amarande)
[Jon Douglas](https://www.linkedin.com/in/jon-douglas-9555572b)
[Liam Moat](https://www.linkedin.com/in/liammoatcom)
[Melanie Guittet](https://www.linkedin.com/in/melanie-guittet/)

```

![pinned-dependencies](./Scorecard-Pinned-Dependencies.png)
![packaging](./Scorecard-Packaging.png)
 
Other checks illustrated below, which are not tied to specific ecosystems and rely solely on GitHub APIs, highlight low-hanging fruit opportunities for .NET ecosystem maintainers to make improvements and enhance open-source security. Some of them like having branch protection mechanisms in place, have only read permissions for GH tokens, SAST (usage of static code analysis tools) and do code reviews would be easy to implement and have a big impact on the overall security of the repository.
 
![branch-protection](./Scorecard-Branch-Protection.png)
![token-permissions](./Scorecard-Token-Permissions.png)
![sast](./Scorecard-SAST.png)
![code-reviews](./Scorecard-Code-Reviews.png)
![mantained](./Scorecard-Mantained.png)
![signed-release](./Scorecard-Signed-Release.png)
![vulnerabilities](./Scorecard-Vulnerabilities.png)
![binary-artifacts](./Scorecard-Binary-Artifacts.png)
 
# Recommendations and action items

> With the recent improvements in handling pinned dependencies and packaging checks for the NuGet package manager, the ability to assess the security features of GitHub repositories in the .NET ecosystem will be substantially enhanced. This should encourage repositories’ maintainers to adopt necessary improvements, leading to higher scores measured by Scorecard and improved overall security.

Reviewer 2: ```suggestion
With the recent improvements in handling pinned dependencies and packaging checks for the NuGet package manager, the ability to assess the security features of GitHub repositories in the .NET ecosystem will be substantially enhanced. This should encourage repositories’ maintainers to adopt necessary improvements, leading to higher scores measured by Scorecard and improved overall security of the entire .Net ecosystem.
```

In conclusion, regularly running OSSF Scorecard checks helps ensure your project stays secure, up-to-date, and aligned with coding best practices. This proactive approach significantly reduces the risk of security vulnerabilities within your software ecosystem.

> Don’t delay! Check the Scorecard of your favorite GitHub repository today and if you’d like to improve the score, take a look at the suggested mitigation steps 😊.

Reviewer 2: ```suggestion
Don’t delay! Check the Scorecard of your or your dependency's GitHub repository today and if you’d like to improve the score, take a look at the suggested mitigation steps 😊. Also, you may further contribute to the Scorecard support of the .Net ecosystem by checkout out the future work section next.
```

# References and future work

> - OpenSSF [Scorecard repository](https://github.com/ossf/scorecard)
> 
> The following links capture open issues for extending Scorecard support:

Reviewer 2: ```suggestion
The following links capture open issues for extending the [OpenSSF Scorecard](https://github.com/ossf/scorecard) support for the .Net and Nuget ecosystem.

```
